internal/proxy/proxy.go

```diff
@@ -133,32 +143,76 @@ func NewClient(ctx context.Context, d cloudsql.Dialer, cmd *cobra.Command, conf
 		}(inst.Name)
 	}
 	pc := newPortConfig(conf.Port)
+
 	for _, inst := range conf.Instances {
-		m := &socketMount{inst: inst.Name}
-		a := conf.Addr
-		if inst.Addr != "" {
-			a = inst.Addr
-		}
 		version, err := d.EngineVersion(ctx, inst.Name)
 		if err != nil {
 			return nil, err
 		}
-		var np int
-		switch {
-		case inst.Port != 0:
-			np = inst.Port
-		case conf.Port != 0:
-			np = pc.nextPort()
-		default:
-			np = pc.nextDBPort(version)
+
+		var (
+			// network is one of "tcp" or "unix"
+			network string
+			// address is either a TCP host port, or a Unix socket
+			address string
+		)
+		if (conf.UnixSocket != "" || inst.UnixSocket != "") &&
```

Comment:
nit: I think maybe we should invert this, and do TCP first since that's the "default" experience. 

it would also be nice to have a comment here explaining what we are checking with these 4 conditions.

Reply:
Sounds good. I've added a comment as well. We might be able to tighten up this boolean expression as well.